### ISSUE TITLE
injecting an error handler into a test

### ISSUE THREAD

**Issue by mentaal** [pytest-dev/pytest#7267]
Hi there.
I have a common error handler that I'd like to run when a subset of my tests fail. Say, when the ones using a particular fixture fail. I'd to do inject this somehow without having to explicitly reference it in the tests. 

What would be great is if I could do something like this:

```python

@pytest.fixture
def error_handler(some_fixture):
    try:
        yield
    except SomeException as e:
        print_extra_params()
```

I know this isn't possible because error exceptions are not thrown back into the tear down code. 

How else could I do something like this in pytest? Or to rephrase, in pytest, how can I inject an error handler into a test such that I don't have to explicitly reference it in a test. 

At present I have been doing this with a custom `pytest_pyfunc_call`. This works fine except that I cannot use fixtures within the function. So it doesn't feel like the correct way to approach this problem. 

Like this:

```python
@pytest.hookimpl(hookwrapper=True)
def pytest_pyfunc_call(pyfuncitem):

    funcargs = pyfuncitem.funcargs
    testargs = {arg: funcargs[arg] for arg in pyfuncitem._fixtureinfo.argnames}

    to_check = {'fixture1', 'fixture2', 'fixture3'}
    common = to_check & testargs.keys()
    if common:
        key = common.pop()
        api = testargs[key]
        try:
            outcome = yield
            outcome.get_result()
        except SomeException as e:
            extra_checks()
            raise
    else:
        outcome = yield
        outcome.get_result()
```

I can understand why in most cases it wouldn't be desirable to have  teardown code have to care about catching errors. I wonder if it would be feasible to add it as an optional change in behavior though, such as with a keyword argument for fixture for argument's sake.

**Comment by RonnyPfannschmidt** [comment]
typically the makereport hook is used for that, as it happens before teadown is executed, is there any reason you cant use a markereport hook?

**Comment by mentaal** [comment]
Hi Ronny, 
I see, it does look more suitable than `pytest_pyfunc_call`. 
Though like pyfunc_call, I still can't use a fixture from there.... right?
It would be great if the error_handler function could also leverage the power and flexibility of fixtures. 
Otherwise, I have to duplicate the logic I have already encoded within the fixtures for this purpose. 
Am I correct? Is there any way to use them at this level?

**Comment by RonnyPfannschmidt** [comment]
@mentaal ye, thats currently not exposed, fixtures are currently intentionally isolated from test failures/exceptions

if you need current item extra metadata its strongy suggested to have a integrated fixture, that can stick it to the self of a plugin object while under test


a crude example would be something like (untested):
```python
class PluginHelper:
   extra_data = None
   @pytest.fixture
   def funstuff(self, request):
      self.extra_data = omg
      yield something
      self.extra_data = None

  def pytest_runtest_makereport(self, when, call):
     if when != "call":
      return
     report = (yield).get_outcome()
     if not report.passed:
         report.sections.append("extra_data", str(self.extra_data)

**Comment by mentaal** [comment]
Got it thanks. 

In order to get at the fixtures that I was interested in, I've resorted to an ugly hack but it allows to me utilize session scoped fixtures within a pytest_pyfunc_call.

I created a fixture with `autouse=True` which has session scoped dependency fixtures which I am actually interested in.

Then I have a helper function:

```python
def get_fixture(pyfuncitem, fixture_name):
    return pyfuncitem._fixtureinfo.name2fixturedefs[fixture_name][0].cached_result[0]
```
Which I use within my `pytest_pyfunc_call` definition to actually get at the fixtures.
Its ugly but it works.
